### PR TITLE
Commercial: add resize message

### PR DESCRIPTION
**Pull request by regiskuckaertz**
## What does this change?

Next task in the pipe for our cross-frame communication protocol, is to allow the ad to ask the page to resize the slot to a specific size. In this iteration, we will deliberately resize the ad slot as specified, but in the future we might adopt an approach ala AMP, where a user has to advertise intent (e.g. click on a button) for the ad to get expanded.
## What is the value of this and can you measure success?

For all our slots, we will try to set a fixed height according to the type of creative we expect back from DFP. But this size might not always be ideal across all breakpoints and creatives, so there is a need for dynamic resizing.
## Does this affect other platforms - Amp, Apps, etc?

Nope, not yet.

cc @guardian/commercial-dev

### CODE REVIEW

File: static/src/javascripts/projects/common/utils/closest.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+define(function () {
+    /**
+     * Adds a wrapper around Element.closest()
+     */
+    var matches = 'matches' in Element.prototype ? 'matches' : 'msMatchesSelector';
```
Comment (rich-nguyen) [inline static/src/javascripts/projects/common/utils/closest.js]: are we sure that all polyfilled browsers have either `matches` or `msMatchesSelector` in `Element` proto?

Comment (regiskuckaertz) [inline static/src/javascripts/projects/common/utils/closest.js]: I based this on the assumption that `Element.matches` exists for all browsers that are running commercial, except IE9 which uses `msMatchesSelector`. I may need to add the `webkit` prefix too [for older androids](http://caniuse.com/#feat=matchesselector) though. @ScottPainterGNM do you agree?

File: static/src/javascripts/projects/common/utils/closest.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    return 'closest' in Element.prototype ? closestNative : closestPolyfill;
+
+    function closestNative(element, selectors) {
+        return element.closest(selectors);
+    }
+
+    function closestPolyfill(element, selectors) {
+        while (element && !element[matches](selectors)) {
```
Comment (rich-nguyen) [inline static/src/javascripts/projects/common/utils/closest.js]: I think (not sure!) there is a bug here. on the first while iteration, we are running `matches` on the specified node, not the first ancestor. So if the specified node happens to contain descendant nodes that match the selector, the polyfill will return the specified node, even when no closest ancestor satisfies the selector.

File: static/src/javascripts/projects/common/utils/closest.js (hunk 1 above)
Comment (rich-nguyen) [inline static/src/javascripts/projects/common/utils/closest.js]: Polyfill for `matches` here: https://developer.mozilla.org/en-US/docs/Web/API/Element/matches

File: static/src/javascripts/projects/common/utils/closest.js (hunk 2 above)
Comment (regiskuckaertz) [inline static/src/javascripts/projects/common/utils/closest.js]: This is indeed counter-intuitive, but it is [how the `closest` method works](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest). ECMAScript authors would be lambasted by taxonomists, that's for sure 😆

Comment (rich-nguyen) [inline static/src/javascripts/projects/common/utils/closest.js]: Just tested it, it's fine, ignore the above. I thought `matches` was using a context, but it doesn't.

**Comment by rich-nguyen** [comment]
👍  I wish there was a clear way to distinguish amd modules which are used for module initialisation (like registering handlers), and ones which return usable module objects.

File: common/app/templates/inlineJS/blocking/polyfills/matches.scala.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+@()
+
+/* Source: https://github.com/remy/polyfills/blob/master/classList.js */
```
Comment (rich-nguyen) [inline common/app/templates/inlineJS/blocking/polyfills/matches.scala.js]: Is that the right source url?

**Comment by rich-nguyen** [comment]
@regiskuckaertz do you think we should have some sort of convention for modules like `get-stylesheet` and `resize`? Otherwise it will soon be hard to see the messages available to be used in the protocol. Similarly for the `set-ad-height` message, i'm not sure where the best place is for that. it should probably be changed to use `resize` ?

File: static/src/javascripts/projects/common/utils/closest.js (hunk 1 above)
Comment (ScottPainterGNM) [inline static/src/javascripts/projects/common/utils/closest.js]: Sorry I was late to the party on this but yes I think it's worth adding the `webkit` prefix as the uptake of new versions of Android is way slower than iOS and as such we can expect a fair amount of traffic across versions of Android 4.

**Comment by regiskuckaertz** [comment]
@rich-nguyen 

- re AMD modules: indeed this distinction does not really exist and the problem won't go away as we switch to ES6 modules. We could come up with some nomenclature and also try to segregate the ones from the others.

- re convention: you mean, in terms of their internal structure? or naming? the `set-ad-height` is quite problematic, it is intricately related to the sticky banner; and the fact that there are many duplicate rubicon line items make it hard to update it. I can probably pipe it to `resize` though, very good idea!

**Comment by rich-nguyen** [comment]
@regiskuckaertz I was thinking any simple convention which distinguishes modules which are protocol-message-listeners from the modules that make up the functional parts of the decomposed old `dfp-api`. Think this would help readers.

- Yes, maybe move `messenger.js`, and message-listener modules into a sub folder within `private`, like `cross-frame-messages`.
- Just naming. So, always names modules that listen to messages by the message name, eg. `get-stylesheet.js`  would be `get-styles.js`.

Piping to `resize` seems sensible, it feels like `set-ad-height` was the first instance of cross-frame comms, so now we have a general implementation (`resize`), they should work through one shared message handler, allowing us to safely migrate away from `set-ad-height` within the DFP scripts.

**Comment by johnduffell** [comment]
conflicts! 😢 is this ready to go?

**Comment by regiskuckaertz** [comment]
Closing this until the DFP team figures out a solution to the problem I've reported.

**Pull request closed without merging**